File: ObjectiveC/FlexChart101/FlexChart101/HitTestController.m
//
//  HitTestController.m
//  FlexChart101
//
//  Copyright (c) 2015 GrapeCity. All rights reserved.
//

#import "HitTestController.h"
#import "HitTestData.h"
@import XuniFlexChartDynamicKit;

@interface HitTestController ()
@property (weak, nonatomic) IBOutlet FlexChart *chart;
@property (weak, nonatomic) IBOutlet UILabel *chartElementLabel;
@property (weak, nonatomic) IBOutlet UILabel *seriesLabel;
@property (weak, nonatomic) IBOutlet UILabel *pointIndexLabel;
@property (weak, nonatomic) IBOutlet UILabel *xyValuesLabel;

@end

@implementation HitTestController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    
    FlexChart *chart = self.chart;
    chart.bindingX = @"x";
    XuniSeries *seriesCosX = [[XuniSeries alloc] initForChart:chart binding:@"y, y" name:@"cos(x)"];
    XuniSeries *seriesSinX = [[XuniSeries alloc] initForChart:chart binding:@"y, y" name:@"sin(x)"];
    seriesCosX.itemsSource = [HitTestData cosData];
    seriesSinX.itemsSource = [HitTestData sinData];
    
    [chart.series addObject:seriesCosX];
    [chart.series addObject:seriesSinX];
    
    chart.delegate = self;
    chart.chartType = XuniChartTypeLineSymbols;
    chart.axisY.format = @"F";
    chart.header = @"Trigonometric Functions";
    chart.footer = @"Cartesian coordinates";

}



-(BOOL)tapped:(FlexChartBase *)sender point:(XuniPoint *)point
{
    FlexChart *chart = self.chart;
    UILabel *s = self.seriesLabel;
    UILabel *pi = self.pointIndexLabel;
    UILabel *xy = self.xyValuesLabel;
    UILabel *ce = self.chartElementLabel;
    XuniChartHitTestInfo *hitTest = [chart hitTest:point];
    pi.text = [@" Point Index: " stringByAppendingString: [NSString stringWithFormat:@"%i", hitTest.dataPoint.pointIndex]];
    xy.text = [[NSString stringWithFormat:@" X:%@", hitTest.dataPoint.valueX] stringByAppendingString: [NSString stringWithFormat:@" Y:%1.2f", hitTest.dataPoint.value]];
    if (hitTest.dataPoint.seriesIndex >= 0 && hitTest.dataPoint.seriesIndex < chart.series.count) {
        s.text = [@" Series: " stringByAppendingString: hitTest.dataPoint.seriesName];
    }
    else
    {
        s.text = @" Series: ";
    }
    ce.text = [@" Chart element: " stringByAppendingString: [self getChartTypeString:(hitTest.chartElement)]];
    
    return false;
}

-(NSString *)getChartTypeString: (XuniChartElement) chartElement {
    switch (chartElement) {
        case XuniChartElementPlotArea:
            return @"PlotArea";
            break;
        case XuniChartElementAxisX:
            return @"AxisX";
            break;
        case XuniChartElementAxisY:
            return @"AxisY";
            break;
        case XuniChartElementChartArea:
            return @"ChartArea";
            break;
        case XuniChartElementLegend:
            return @"Legend";
            break;
        case XuniChartElementHeader:
            return @"Header";
            break;
        case XuniChartElementFooter:
            return @"Footer";
            break;
        case XuniChartElementNone:
            return @"None";
            break;
        default:
            return @"Invalid Type";
            break;
    }
}


@end

File: Swift/FlexChart101/FlexChart101/HitTestController.swift
//
//  HitTestController.swift
//  FlexChart101
//
//  Copyright (c) 2015 GrapeCity. All rights reserved.
//

import UIKit
import XuniFlexChartKit

class HitTestController: UIViewController, FlexChartDelegate {
    
    var _chart = FlexChart()
    var _chartElementLabel = UILabel()
    var _seriesLabel = UILabel()
    var _pointIndexLabel = UILabel()
    var _xyLabel = UILabel()

    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = NSLocalizedString("Hit Test", comment: "")
        
        // Do any additional setup after loading the view.
        _chartElementLabel.text = " Chart element"
        _seriesLabel.text = " Series"
        _pointIndexLabel.text = " Point Index"
        _xyLabel.text = " X Y Values"
        
        let seriesCosX = XuniSeries(forChart: _chart, binding: "y, y", name: "cos(x)")
        let seriesSinX = XuniSeries(forChart: _chart, binding: "y, y", name: "sin(x)")
        
        seriesCosX.itemsSource = HitTestData.cosData()
        seriesSinX.itemsSource = HitTestData.sinData()
        
        _chart.series.addObject(seriesCosX)
        _chart.series.addObject(seriesSinX)
        
        _chart.bindingX = "x"
        _chart.delegate = self
        _chart.chartType = XuniChartType.LineSymbols
        _chart.axisY.format = "F"
        _chart.header = "Trigonometric Functions"
        _chart.footer = "Cartesian coordinates"
        
        self.view.addSubview(_chart)
        self.view.addSubview(_chartElementLabel)
        self.view.addSubview(_seriesLabel)
        self.view.addSubview(_pointIndexLabel)
        self.view.addSubview(_xyLabel)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        _chart.frame = CGRectMake(0, self.view.bounds.size.height / 8, self.view.bounds.size.width, self.view.bounds.size.height * 5 / 8)
        _seriesLabel.frame = CGRectMake(0, self.view.bounds.size.height * 13 / 16, self.view.bounds.size.width, self.view.bounds.size.height / 16)
        _pointIndexLabel.frame = CGRectMake(0, self.view.bounds.size.height * 14 / 16, self.view.bounds.size.width, self.view.bounds.size.height / 16)
        _xyLabel.frame = CGRectMake(0, self.view.bounds.size.height * 15 / 16, self.view.bounds.size.width, self.view.bounds.size.height / 16)
        _chartElementLabel.frame = CGRectMake(0, self.view.bounds.size.height * 6 / 8, self.view.bounds.size.width, self.view.bounds.size.height / 16)
    }
    
    func tapped(sender: XuniView!, point: XuniPoint!) -> Bool {
        let hitTest = _chart.hitTest(point)
        var seriesName = "";
        if (hitTest.dataPoint.seriesIndex >= 0 && hitTest.dataPoint.seriesIndex < Int32(_chart.series.count)) {
            seriesName = hitTest.dataPoint.seriesName;
        }
        
        _pointIndexLabel.text = " Point Index: " + String(hitTest.dataPoint.pointIndex);
        _xyLabel.text = " X:" + String(format: "%@", hitTest.dataPoint.valueX ?? "0") + " Y:" + String(format: "%1.2f", hitTest.dataPoint.value)
        _seriesLabel.text = " Series: " + seriesName
        _chartElementLabel.text = " Chart element: " + getChartTypeString(hitTest.chartElement)
        
        return false;
    }
    
    func getChartTypeString(chartElement: XuniChartElement) -> String {
        switch (chartElement) {
            case XuniChartElement.PlotArea:
                return "PlotArea"
            case XuniChartElement.AxisX:
                return "AxisX"
            case XuniChartElement.AxisY:
                return "AxisY"
            case XuniChartElement.ChartArea:
                return "ChartArea"
            case XuniChartElement.Legend:
                return "Legend"
            case XuniChartElement.Header:
                return "Header"
            case XuniChartElement.Footer:
                return "Footer"
            case XuniChartElement.None:
                return "None"
            default:
                return "Invalid Type"
        }
    }
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

}
